File: Algorithms/Medium/ContainerWithMostWater.go
package medium

// Time complexity: O(n)
// Space complexity: O(1)
func maxArea(height []int) int {
	l, r := 0, len(height)-1
	max_area := 0

	for l < r {
		area := (r - l) * min(height[l], height[r])
		max_area = max(max_area, area)
		if height[l] < height[r] {
			l++
		} else {
			r--
		}
	}

	return max_area
}

File: Algorithms/Medium/11.Container-With-Most-Water/ContainerWithMostWater.py
# Time complexity: O(n)
# Space complexity: O(1)
class Solution:
    def maxArea(self, height: list[int]) -> int:
        max_area = 0
        l, r = 0, len(height) - 1
        while l < r:
            area = min(height[l], height[r]) * (r - l)
            if area > max_area:
                max_area = area
            
            if height[l] < height[r]:
                l += 1
            else:
                r -= 1
        
        return max_area